fixes for julia 0.7

fredrikekre opened the pull request.

ararslan commented:
I'm not sure we want to drop support for 0.6 here yet.

fredrikekre commented:
Why? If it turns out a new release compatible with julia 0.6 is needed we can do that afterwards from a branch.

ararslan commented:
Yeah, I guess. It's @simonster's call.

ararslan commented on test/recon.jl:
```diff
@@ -1,321 +0,0 @@
-using JLD2
```
Perhaps using Revise?

fredrikekre replied:
Ah, I misunderstood the comment here https://github.com/simonster/JLD2.jl/pull/79/files#diff-fce720c43af3c52c862fd7451c7374b8R12 and thought these tests were specific to `workspace`.

ararslan replied:
No, they just require that kind of functionality.

fredrikekre commented:
Updated; I fixed the reconstruction tests by spawning another process that defines and saves the custrom structs. Then the test-process loads and reconstructs them.

fredrikekre commented:
Hmm, Windows error:
```
MethodError: Cannot `convert` an object of type Base.Libc.WindowsRawSocket to an object of type Int64
Closest candidates are:
  convert(::Type{T<:Number}, !Matched::T<:Number) where T<:Number at number.jl:6
  convert(::Type{T<:Number}, !Matched::Number) where T<:Number at number.jl:7
  convert(::Type{T<:Integer}, !Matched::Ptr) where T<:Integer at pointer.jl:23
  ...
Stacktrace:
 [1] cconvert(::Type, ::Base.Libc.WindowsRawSocket) at .\essentials.jl:328
 [2] mmap! at C:\projects\jld2-jl\src\mmapio.jl:61 [inlined]
 [3] JLD2.MmapIO(::String, ::Bool, ::Bool, ::Bool) at C:\projects\jld2-jl\src\mmapio.jl:113
 [4] top-level scope
 [5] include at .\boot.jl:314 [inlined]
 [6] include_relative(::Module, ::String) at .\loading.jl:1071
 [7] include(::Module, ::String) at .\sysimg.jl:29
 [8] include(::String) at .\client.jl:393
 [9] top-level scope
 [10] include at .\boot.jl:314 [inlined]
 [11] include_relative(::Module, ::String) at .\loading.jl:1071
 [12] include(::Module, ::String) at .\sysimg.jl:29
 [13] include(::String) at .\client.jl:393
 [14] top-level scope
```
Not exactly sure what to do about that, but happens on JLD2 master also. The error is in this `ccall`:
```julia
        mapping = ccall(:CreateFileMappingW, stdcall, Ptr{Cvoid},
                        (Cptrdiff_t, Ptr{Cvoid}, DWORD, DWORD, DWORD, Ptr{Cvoid}),
                        Mmap.gethandle(io.f), C_NULL,
                        io.write ? Mmap.PAGE_READWRITE : Mmap.PAGE_READONLY, n >> 32,
                        n % UInt32, C_NULL)
```
Related to some change in `Mmap` for 0.7?

alandion commented:
> Related to some change in Mmap for 0.7?
`Base.Libc.WindowsRawSocket` is a primitive type now in 0.7 and no convert to `Cptrdiff_t` is defined.  A possible fix is to replace

`Mmap.gethandle(io.f)`

with 

`Base.cconvert(Ptr{Cvoid},Mmap.gethandle(io.f))`

fredrikekre commented:
It used to pass tests on Linux, but now there is a new error:
```
e = InexactError(:check_top_bit, UInt64, 0xc2a01210000081e7)
Stacktrace:
 [1] throw_inexacterror(::Symbol, ::Any, ::UInt64) at ./boot.jl:567
 [2] check_top_bit at ./boot.jl:581 [inlined]
 [3] toInt64 at ./boot.jl:642 [inlined]
 [4] Type at ./boot.jl:717 [inlined]
 [5] fileoffset at /home/travis/build/simonster/JLD2.jl/src/JLD2.jl:180 [inlined]
 [6] read_heap_object(::JLD2.JLDFile{IOStream}, ::JLD2.GlobalHeapID, ::JLD2.ReadRepresentation{UInt8,UInt8}) at /home/travis/build/simonster/JLD2.jl/src/global_heaps.jl:127
 [7] jlconvert at /home/travis/build/simonster/JLD2.jl/src/data.jl:701 [inlined]
 [8] jlconvert(::JLD2.ReadRepresentation{DataType,JLD2.OnDiskRepresentation{(0, 16),Tuple{String,Array{Any,1}},Tuple{JLD2.Vlen{String},JLD2.Vlen{JLD2.RelOffset}}}()}, ::JLD2.JLDFile{IOStream}, ::Ptr{UInt8}, ::JLD2.RelOffset) at /home/travis/build/simonster/JLD2.jl/src/data.jl:908
 [9] read_scalar at /home/travis/build/simonster/JLD2.jl/src/dataio.jl:196 [inlined]
 [10] read_data(::JLD2.JLDFile{IOStream}, ::JLD2.ReadRepresentation{DataType,JLD2.OnDiskRepresentation{(0, 16),Tuple{String,Array{Any,1}},Tuple{JLD2.Vlen{String},JLD2.Vlen{JLD2.RelOffset}}}()}, ::Nothing) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:171
 [11] read_data(::JLD2.JLDFile{IOStream}, ::JLD2.ReadDataspace, ::UInt8, ::Int64, ::Int64, ::Int64, ::UInt16, ::JLD2.RelOffset, ::Nothing) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:149
 [12] read_data(::JLD2.JLDFile{IOStream}, ::JLD2.ReadDataspace, ::UInt8, ::Int64, ::Int64, ::Int64, ::UInt16) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:144 (repeats 3 times)
 [13] read_attr_data(::JLD2.JLDFile{IOStream}, ::JLD2.ReadAttribute) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:102
 [14] jltype(::JLD2.JLDFile{IOStream}, ::JLD2.CommittedDatatype) at /home/travis/build/simonster/JLD2.jl/src/data.jl:306
 [15] read_data(::JLD2.JLDFile{IOStream}, ::JLD2.ReadDataspace, ::UInt8, ::Int64, ::Int64, ::Int64, ::UInt16, ::JLD2.RelOffset, ::Array{JLD2.ReadAttribute,1}) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:146
 [16] load_dataset(::JLD2.JLDFile{IOStream}, ::JLD2.RelOffset) at /home/travis/build/simonster/JLD2.jl/src/datasets.jl:92
 [17] getindex(::JLD2.Group{JLD2.JLDFile{IOStream}}, ::String) at /home/travis/build/simonster/JLD2.jl/src/groups.jl:108
 [18] read(::JLD2.JLDFile{IOStream}, ::String) at /home/travis/build/simonster/JLD2.jl/src/JLD2.jl:326
 [19] top-level scope at /home/travis/build/simonster/JLD2.jl/test/rw.jl:379
```
which happens locally for me as well. I am not sure what changed, and what needs to be fixed.

staticfloat commented:
Thanks for the work here @fredrikekre; I actually didn't see this when I was doing my own updates, but I'm glad we largely converged on our results.  :P

fredrikekre commented:
I'll revive this with the fixes to `test/recon.jl` which have not yet been fixed.
